app/components/user/TwoFactor/RecoveryCodes/RecoveryCodeDialog.js

```diff
@@ -0,0 +1,206 @@
+  if (recoveryCodes && recoveryCodes.codes) {
+    return recoveryCodes.codes.reduce((memo, { code }) => memo.concat(code), []).join('\n');
+  }
+  return '';
+}
+
+class RecoveryCodesDialog extends React.PureComponent<Props, State> {
+  state = {
+    generatingRecoveryCodes: false,
+    generatedRecoveryCodes: false,
+    recoveryCodes: null
+  };
+
+  componentDidMount() {
+    // If we have a TOTP, but no recovery codes for it, immediately go generate some recovery codes.
+    if (this.props.totp && !this.props.totp.recoveryCodes) {
+      this.generateRecoveryCodes();
+    }
+  }
+
+  render() {
+    return (
+      <Dialog isOpen={true} onRequestClose={this.props.onRequestClose} width={540}>
+        <div className="p4">
+          {(this.props.totp.recoveryCodes
+            ? this.renderTotpRecoveryCodes()
+            : this.renderGenerateTotpRecoveryCodes()
+          )}
+        </div>
+      </Dialog>
+    );
+  }
+
+  // TODO: Dry this up...
+  renderGenerateTotpRecoveryCodes() {
+    if (this.state.generatingRecoveryCodes) {
+      return (
+        <React.Fragment>
+          <h2 className="m0 h2 semi-bold mb3">Generating Recovery Codes</h2>
+          <p>Please wait, weʼre getting some recovery codes ready for you.</p>
+        </React.Fragment>
+      );
+    }
+
+    return (
+      <React.Fragment>
+        <h2 className="m0 h2 semi-bold mb3">Generated Recovery Codes</h2>
+        <p>Recovery codes are used if you lose access to your code generator.</p>
+        <RecoveryCodeList recoveryCodes={this.state.recoveryCodes} />
+        <p>These codes should be treated just like your password. Weʼd suggest saving them into a secure password manager.</p>
+        <CopyToClipboard
+          text={recoveryCodeText(this.state.recoveryCodes)}
+          onCopy={this.handleRecoveryCodeCopy}
+        >
+          <Button
+            className="col-12"
+            theme="success"
+            outline={true}
+          >
+            Copy Recovery Codes
+          </Button>
+        </CopyToClipboard>
+      </React.Fragment>
+    );
+  }
+
+  // TODO: Dry this up...
+  renderTotpRecoveryCodes() {
+    return (
+      <React.Fragment>
+        <h2 className="m0 h2 semi-bold mb5">Current Recovery Codes</h2>
+        <p>Recovery codes are used if you lose access to your code generator.</p>
+        <RecoveryCodeList recoveryCodes={this.props.totp.recoveryCodes} />
+        <p>These codes should be treated just like your password. Weʼd suggest saving them into a secure password manager.</p>
+        <CopyToClipboard
+          text={recoveryCodeText(this.props.totp.recoveryCodes)}
+          onCopy={this.handleRecoveryCodeCopy}
+        >
+          <Button
+            className="col-12"
+            theme="success"
+            outline={true}
+          >
+            Copy Recovery Codes
+          </Button>
+        </CopyToClipboard>
+      </React.Fragment>
+    );
+  }
+
+  generateRecoveryCodes() {
```

Comment:
Might be worth pulling this whole function & mutation up into the parent component...?